src/barril/units/_array.py

```diff
@@ -178,6 +181,51 @@ def _DoValidateValues(self, values, quantity):
                             # Break the outer 'for' used just to get the min/max
                             break
 
+    @classmethod
+    def FromScalars(
+        cls,
+        scalars: Iterable[Scalar],
+        *,
+        unit: Optional[str] = None,
+        category: Optional[str] = None
+    ):
+        """
+        Create an Array from a sequence of Scalars.
+
+        When not defined, the unit and category assumed will be from the first Scalar on the sequence.
+
+        :param values:
+            A sequence of Scalars. When the values parameter is an empty sequence and
+            the unit is not provided an Array with an empty quantity will be returned.
+
+        :param unit:
+            A string representing the unit, if not defined
+            the unit from the first Scalar on the sequence will be used.
+
+        :param category:
+            A string representing the category, if not defined
+            the category from the first Scalar on the sequence will be used.
+        """
+        scalars = iter(scalars)
+        try:
+            first_scalar = next(scalars)
+        except StopIteration:
```

Comment:
Why all these iterators juggling?

IMHO, a more "traditional" approach would be a lot more legible:

```
if len(scalars) == 0:
    ...
    return

first_scalar = scalars[0]
...
values = [scalar.GetValue(unit) for scalar in scalars]
...

Reply:
We wanted to receive an iterator as argument, and the usual approach when dealing with iterators is to call `iter()` on the object right away (to get a real iterator, validate against non-iterables, etc).

So you can't call `len` on iterators:

```python
>>> it = iter([1, 2, 3])
>>> len(it)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'list_iterator' has no len()
```

Checking `if scalars` is also a no-go:

```python
>>> bool(it)
True
>>>
```

So the usual approach is use `next` explicitly.

Using an iterator here has the advantage of supporting generators when calling `Array.FromScalars` with generators.

Also IMO the real ugly part here is handling all 3 cases regarding `unit` and `category`, but that's a general flaw in how we designed `barril` around explicit `unit` and `category` parameters.